src/utransport.rs

```diff
@@ -27,37 +27,53 @@ use zenoh::{
 };
 
 #[inline]
-fn error_callback(
-    sender: &Sender<CallbackChannelMessage>,
-    listener: &Arc<dyn UListener>,
-    msg: &str,
-) {
-    log::error!("{msg}");
-    if sender
-        .send(CallbackChannelMessage {
-            listener: listener.clone(),
-            result: Err(UStatus::fail_with_code(UCode::INTERNAL, msg)),
-        })
-        .is_err()
-    {
-        log::error!("Unable to call the user callback");
+fn invoke_block_callback(listener: &Arc<dyn UListener>, resp_msg: Result<UMessage, &str>) {
+    match resp_msg {
+        Ok(umsg) => {
+            task::block_in_place(|| {
+                Handle::current().block_on(listener.on_receive(umsg));
+            });
+        }
+        Err(msg) => {
+            log::error!("{msg}");
+            task::block_in_place(|| {
+                Handle::current()
+                    .block_on(listener.on_error(UStatus::fail_with_code(UCode::INTERNAL, msg)));
+            });
+        }
     }
 }
 
 #[inline]
-fn success_callback(
+fn invoke_nonblock_callback(
     sender: &Sender<CallbackChannelMessage>,
     listener: &Arc<dyn UListener>,
-    msg: UMessage,
+    listener_msg: Result<UMessage, &str>,
 ) {
-    if sender
-        .send(CallbackChannelMessage {
-            listener: listener.clone(),
-            result: Ok(msg),
-        })
-        .is_err()
-    {
-        log::error!("Unable to call the user callback");
+    match listener_msg {
+        Ok(umsg) => {
+            if sender
+                .send(CallbackChannelMessage {
+                    listener: listener.clone(),
+                    result: Ok(umsg),
+                })
+                .is_err()
+            {
+                log::error!("Unable to call the user callback");
+            }
+        }
+        Err(msg) => {
+            log::error!("{msg}");
+            if sender
+                .send(CallbackChannelMessage {
+                    listener: listener.clone(),
+                    result: Err(UStatus::fail_with_code(UCode::INTERNAL, msg)),
```

Comment:
What about [`RESOURCE_EXHAUSTED`](https://github.com/eclipse-uprotocol/up-spec/blob/19bec77b3c3222def74604dbe5f5d0af2afb9ae1/up-core-api/uprotocol/ustatus.proto#L84) instead? Seems more fitting to me